### ISSUE TITLE
Bug: Error when using pnpm with package.yaml, 'type: module' not recognized

### ISSUE THREAD

**Issue by igorakkerman** [typescript-eslint/typescript-eslint#8889]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

typescript-eslint

### Repro Code

```TypeScript
Small TypeScript / PNPM project to reproduce the issue:
https://github.com/igorakkerman/eslint-package-yaml-demo

The project contains a package.yaml but no package.json.
```

### Expected Result

ESLint should recognize that pnpm uses *package.yaml* which contains `type: module` and treat the project as such.

### Actual Result

```
> pnpm eslint
(node:14516) Warning: To load an ES module, set "type": "module" in the package.json or use the .mjs extension.
(Use `node --trace-warnings ...` to show where the warning was created)

Oops! Something went wrong! :(

ESLint: 9.0.0

.../eslint-package-yaml-demo/eslint.config.js:1
import typescriptParser from '@typescript-eslint/parser';
^^^^^^

SyntaxError: Cannot use import statement outside a module
    at internalCompileFunction (node:internal/vm:128:18)
    at wrapSafe (node:internal/modules/cjs/loader:1279:20)
    at Module._compile (node:internal/modules/cjs/loader:1331:27)
    at Module._extensions..js (node:internal/modules/cjs/loader:1426:10)
    at Module.load (node:internal/modules/cjs/loader:1205:32)
    at Module._load (node:internal/modules/cjs/loader:1021:12)
    at cjsLoader (node:internal/modules/esm/translators:366:17)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:315:7)
    at ModuleJob.run (node:internal/modules/esm/module_job:222:25)
    at async ModuleLoader.import (node:internal/modules/esm/loader:323:24)
```

### Additional Info

Apparently, ESLint looks for *package.json* but not for *package.yaml*.

Workarounds:
- convert *package.yaml* into *package.json*
- rename *eslint.config.js* as *eslint.config.mjs* (change filename extension)

### Versions

Legend: production dependency, optional only, dev only

eslint-package-yaml-demo@0.0.1 .../eslint-package-yaml-demo

devDependencies:
eslint 9.0.0
eslint-config-prettier 9.1.0
prettier 3.2.5
rimraf 5.0.5
typescript 5.4.4
typescript-eslint 7.6.0

**Comment by JoshuaKGoldberg** [comment]
> `package.yaml`

Well, there's the issue. It's super unusual in the Node ecosystem to use any file other than `package.json` there. We certainly don't support it.

In fact, the errors you're getting are from Node.js, not us!

> ```plaintext
> SyntaxError: Cannot use import statement outside a module
>     at internalCompileFunction (node:internal/vm:128:18)
> ```

**Comment by bradzacher** [comment]
We are not the ESLint project - they live over at https://github.com/eslint/eslint.

Though this stack trace isn't an ESLint error - it's straight from nodejs. So it isn't eslint's fault - Node is erroring because it is not respecting the package.yaml. AFAIK nodejs doesn't support a yaml file - only a json file.

Regardless. This is nothing to do with our project.

**Comment by igorakkerman** [comment]
Thank you, @JoshuaKGoldberg and @bradzacher, for looking into this and for the clarification. I'm aware that you're not ESLint ;-) but didn't know on which layer in the stack the issue occurs and didn't dive deep enough to find out, sorry for that and for addressing the issue in the wrong place. I'm also aware that package.json is the default and tried to use package.yaml for the first time myself. 

I was hoping for pnpm to be the only layer dealing with package.json/yaml. Since tsc and prettier both work fine and eslint-typescript deals with the module system, I got here. 

Do you think it makes sense to recreate this issue in the ESLint repo?

**Comment by bradzacher** [comment]
> they require their config file and rely on Node to figure out whether it is CJS or ESM code. 

Almost correct they `import()` it. But same same. They rely on node to figure it out yeah - so if node isn't picking up your package.yaml then that's a node thing. Node defaults to assuming that `.js` files are cjs if it can't find a package json

An explicit `mjs` extension will circumvent node's assumption.

There's no bug here! Node just doesn't support a yaml file so you need to be explicit to work around that.
